File: variants/dotnet-csharp/Telemetry.cs
// =============================================================================
// Telemetry.cs — Hub de telemetria em tempo real (variante .NET/C#).
//
// Espelha o telemetry_hub.py. Sem broker Kafka, consulta o coordenador (op
// METRICS) a cada segundo e exibe as estatísticas ao vivo (mín/máx/média por
// operação, além de re-replicações e coleta de lixo).
// =============================================================================

using System.Text.Json.Nodes;

namespace Dfs;

public static class Telemetry
{
    public static void Run()
    {
        Console.WriteLine("Hub de telemetria (Ctrl+C para sair). Consultando o coordenador a cada 1s...");
        while (true)
        {
            try
            {
                var r = Protocol.Request(Config.CoordinatorHost, Config.CoordinatorPort, new JsonObject { ["op"] = "METRICS" });
                var line = $"[{DateTime.Now:HH:mm:ss}] arquivos={Protocol.Int(r, "files")} " +
                           $"re-replicações={Protocol.Int(r, "rereplications")} GC={Protocol.Int(r, "gc_deletes")}";
                foreach (var (op, mv) in (JsonObject)r["ops"]!)
                {
                    var m = (JsonObject)mv!;
                    line += $" | {op}: n={Protocol.Int(m, "count")} avg={m["avg_ms"]!.GetValue<double>()}ms " +
                            $"min={m["min_ms"]!.GetValue<double>()}ms max={m["max_ms"]!.GetValue<double>()}ms {Protocol.Long(m, "bytes")}B";
                }
                Console.WriteLine(line);
            }
            catch (Exception e) { Console.WriteLine($"[telemetria] coordenador indisponível: {e.Message}"); }
            Thread.Sleep(1000);
        }
    }
}

File: variants/ruby/telemetry.rb
# frozen_string_literal: true

# =============================================================================
# telemetry.rb — Hub de telemetria em tempo real (variante Ruby).
#
# Espelha o telemetry_hub.py do original. Lá, um consumidor Kafka escutava o
# tópico de métricas; aqui, sem broker, o monitor consulta periodicamente o
# coordenador (op METRICS), que agrega as métricas reportadas pelos nós
# ingress/egress, e exibe estatísticas ao vivo (mín/máx/média por operação, além
# de contadores de re-replicação e de coleta de lixo).
#
# Uso: ruby telemetry.rb   (Ctrl+C para sair)
# =============================================================================

require_relative 'config'
require_relative 'lib/protocol'

$stdout.sync = true
puts 'Hub de telemetria (Ctrl+C para sair). Consultando o coordenador a cada 1s...'
loop do
  begin
    r = DFS::Protocol.request(DFS::Config::COORDINATOR_HOST, DFS::Config::COORDINATOR_PORT, 'op' => 'METRICS')
    ts = Time.now.strftime('%H:%M:%S')
    line = "[#{ts}] arquivos=#{r['files']} re-replicações=#{r['rereplications']} GC=#{r['gc_deletes']}"
    (r['ops'] || {}).each do |op, m|
      line += " | #{op}: n=#{m['count']} avg=#{m['avg_ms']}ms min=#{m['min_ms']}ms max=#{m['max_ms']}ms #{m['bytes']}B"
    end
    puts line
  rescue StandardError => e
    puts "[telemetria] coordenador indisponível: #{e.message}"
  end
  sleep 1
end
